File: market/test/java/ru/yandex/market/ff/util/HibernateQueriesExecutionListener.java
package ru.yandex.market.ff.util;

import java.lang.reflect.Method;
import java.util.stream.Collectors;

import org.springframework.test.context.TestContext;
import org.springframework.test.context.support.AbstractTestExecutionListener;

import ru.yandex.market.ff.util.query.count.JpaQueriesCount;
import ru.yandex.market.ff.util.query.count.QueriesCountInspector;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Слушатель исполняемых JPA запросов
 *
 * @author kotovdv 11/08/2017.
 */
public class HibernateQueriesExecutionListener extends AbstractTestExecutionListener {

    public static final String DESCRIPTION = "Asserting JPA queries count";

    @Override
    public void beforeTestMethod(TestContext testContext) {
        if (testContext == null) {
            return;
        }
        Method testMethod = testContext.getTestMethod();

        JpaQueriesCount jpaQueriesCount = testMethod.getAnnotation(JpaQueriesCount.class);
        if (jpaQueriesCount != null) {
            QueriesCountInspector.reset();
        }
    }

    @Override
    public void afterTestMethod(TestContext testContext) throws Exception {
        if (testContext == null) {
            return;
        }
        Method testMethod = testContext.getTestMethod();

        JpaQueriesCount jpaQueriesCount = testMethod.getAnnotation(JpaQueriesCount.class);
        if (jpaQueriesCount != null) {
            int actual = QueriesCountInspector.getCount();
            int expected = jpaQueriesCount.value();

            if (jpaQueriesCount.isThreshold()) {
                assertThat(actual)
                        .as(DESCRIPTION)
                        .withFailMessage(getErrorMessage(expected))
                        .isLessThanOrEqualTo(expected);
            } else {
                assertThat(actual)
                        .as(DESCRIPTION)
                        .withFailMessage(getErrorMessage(expected))
                        .isEqualTo(expected);
            }
        }
    }

    private static String getErrorMessage(int expectedCount) {
        return String.join("\n",
                "\n",
                "Expected query count: " + expectedCount,
                "Actual query count: " + QueriesCountInspector.getCount(),
                "",
                "Actual query list:",
                QueriesCountInspector.getQueries().stream()
                        .map(HibernateQueriesExecutionListener::buildQueryMessage)
                        .collect(Collectors.joining("\n"))
        );
    }

    private static String buildQueryMessage(QueriesCountInspector.QueryInfo query) {
        return query.getSource() + ":    " + query.getSql();
    }
}

File: market/test/integration/kotlin/util/HibernateQueriesExecutionListener.kt
package ru.yandex.market.logistics.calendaring.util

import org.assertj.core.api.Assertions.assertThat
import org.springframework.test.context.TestContext
import org.springframework.test.context.support.AbstractTestExecutionListener
import ru.yandex.market.logistics.test.integration.jpa.JpaQueriesCount
import java.lang.reflect.Method


class HibernateQueriesExecutionListener : AbstractTestExecutionListener() {

    override fun beforeTestMethod(testContext: TestContext?) {
        if (testContext == null) {
            return
        }
        val testMethod: Method = testContext.testMethod
        val jpaQueriesCount: JpaQueriesCount? = testMethod.getAnnotation(JpaQueriesCount::class.java)
        if (jpaQueriesCount != null) {
            QueriesCountInspector.reset()
        }
    }

    @Throws(Exception::class)
    override fun afterTestMethod(testContext: TestContext?) {
        if (testContext == null) {
            return
        }
        val testMethod: Method = testContext.testMethod
        val jpaQueriesCount: JpaQueriesCount? = testMethod.getAnnotation(JpaQueriesCount::class.java)
        if (jpaQueriesCount != null) {
            val actual: Int = QueriesCountInspector.getCount()
            val expected: Int = jpaQueriesCount.value
            if (jpaQueriesCount.isThreshold) {
                assertThat(actual)
                    .`as`("Asserting JPA queries count")
                    .withFailMessage(getErrorMessage(expected))
                    .isLessThanOrEqualTo(expected)
            } else {
                assertThat(actual)
                    .`as`("Asserting JPA queries count")
                    .withFailMessage(getErrorMessage(expected))
                    .isEqualTo(expected)
            }
        }
    }

    private fun getErrorMessage(expectedCount: Int): String? {
        return java.lang.String.join("\n",
            "\n",
            "Expected query count: $expectedCount",
            "Actual query count: " + QueriesCountInspector.getCount(),
            "",
            "Actual query list:",
            QueriesCountInspector.getQueries().joinToString("\n") { query ->
                query.source + ":    " + query.sql
            }

        )
    }
}
